services/outputhost/kafkaAddresser.go

```diff
@@ -0,0 +1,141 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package outputhost
+
+import (
+	"math"
+	"sync"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cherami-server/common"
+)
+
+// Kafka Addresser store address encoding scheme
+//                    -=-=-
+// t = topic+partition (map lookup)
+// o = offset
+//                     ttttttooooooooooooo
+// math.MaxInt64    =  9223372036854775807 = Max store address
+// divisor          =       10000000000000
+// MaxTP            =  922336 <- Note that it's 1 less, which makes the max offset uniform
+// MaxOffset        =        9999999999999
+// MaxAddress       =  9223369999999999999
+// MaxOffset10YrQPS =                31709 <-- The sustained QPS that would exhaust the max offset in 10 years
+//                    -=-=-
+// https://play.golang.org/p/VVMnGgHWSG
+
+const (
+	kafkaAddresserDivisor    = int64(1e13)
+	kafkaAddresserMaxTP      = int64(math.MaxInt64/kafkaAddresserDivisor) - 1
+	kafkaAddresserMaxOffset  = kafkaAddresserDivisor - 1
+	kafkaAddresserMaxAddress = kafkaAddresserMaxTP*kafkaAddresserDivisor + kafkaAddresserMaxOffset
+)
+
+// topicPartition
+type topicPartition struct {
+	Topic     string
+	Partition int32
+}
+
+type kafkaTopicPartitionAddresser struct {
+	sync.RWMutex
+	tp2i   map[topicPartition]int64
+	i2tp   map[int64]*topicPartition
+	nextTP int64
+}
+
+var kafkaAddresser = kafkaTopicPartitionAddresser{}
+
+// GetStoreAddress converts a given topic, partition, and offset into a store address
+func (k *kafkaTopicPartitionAddresser) GetStoreAddress(tp *topicPartition, offset int64, logFn func() bark.Logger) (address storeHostAddress) {
+	calc := func(tpInt int64, offset int64) storeHostAddress {
+		return storeHostAddress((tpInt * kafkaAddresserDivisor) + offset)
+	}
+
+	if offset > kafkaAddresserMaxOffset {
+		logFn().WithFields(bark.Fields{
+			`module`:    `kafkaAddresser`,
+			`topic`:     tp.Topic,
+			`partition`: tp.Partition,
+			`offset`:    offset,
+		}).Error(`Offset out of Kafka addresser design range`)
+	}
+
+	// Pegging the offset will just mean that we don't commit
+	// larger offsets to Kafka, but shouldn't cause an outage
+	offset = common.MinInt64(offset, kafkaAddresserMaxOffset)
```

Comment:
I have a log line for this, but I'll add a metric when we do the metrics pass on all the Kafka changes.